Make message twig template discoverable

mccrodp opened the pull request:
Cleaner version of original PR: https://github.com/Gizra/message/pull/43

mccrodp committed:
WIP: make message twig template descoverable.

mccrodp committed:
Added preproccess fn: template_preprocess_message and hook: message_theme_suggestions_message

mccrodp committed:
Merge remote-tracking branch 'upstream/update-travis' into add-message-template-hook

mccrodp committed:
Add theme function, template preprocess function and suggestions hook to message.module and #theme index to build array in MessageViewBuilder

RoySegall commented on message.module:
```diff
@@ -285,3 +287,71 @@ function message_config_translation_info_alter(&$info) {
+  $suggestions[] = 'message__' . $message->id() . '__' . $sanitized_view_mode;
+
+  return $suggestions;
+}
```
Need a new line in the end of the file.

RoySegall commented:
When rendering the message like this:

``` php
    $foo = Message::load(1);

    $bar = \Drupal::entityTypeManager()->getViewBuilder('message')->view($foo);
```

and put in the twig file the text foo I don't see it in the renderable array.

mccrodp committed:
Merge remote-tracking branch 'upstream/update-travis' into add-message-template-hook

mccrodp committed:
Fixed langcode index issue #35 and build array generation in MessageViewBuilder and added new MessageTemplateSuggestionsTest

mccrodp committed:
Resolved and merged template conflicts

mccrodp committed:
Updated MessageTemplateSuggestionsTest to use entityTypeManager instead of deprecated entityManager.

mccrodp commented:
@RoySegall I added a new test `MessageTemplateSuggestionsTest` which uses `entityTypeManager` call similar to the below. Now that I have added a fix here for #35 I can verify I now see the render array. This may solve some other open issues in the process.

All tests should be green, so this should be ready / almost ready now, please check! :)

> When rendering the message like this:
> 
>    $foo = Message::load(1);
> 
>    $bar = \Drupal::entityTypeManager()->getViewBuilder('message')->view($foo);
> and put in the twig file the text foo I don't see it in the renderable array.

EDIT: If all is well, I think this means that https://github.com/Gizra/message/issues/32 can be closed now too.

mccrodp committed:
Update message_theme function which fixes content population for template, removed title and label indexes / fields from template preprocess fn

mccrodp commented:
@RoySegall - Template is now working, the content variable is populated and field are displayed as expected. I verified through work on Message UI :)

RoySegall commented:
OK. Can you add another test for verify that when we do 

```
$foo = Message::load(1);
$bar = \Drupal::entityTypeManager()->getViewBuilder('message')->view($foo);
// Use actually the service but you got what i mean :)
render($bar);
```

we get the text wrapped with the div?

mccrodp commented:
Ok, I'll add a new `testMessageTextWrapper()` to `MessageUiTest` that does something like:

```
public function testMessageTextWrapper() {
   $type = 'dummy_message';
    // Create message to be rendered.
    $message_type = $this->createMessageType($type, 'Dummy message', '', array('[message:author:name]'));
    $message = Message::create(array('type' => $message_type->id()))
      ->setAuthorId($this->user->id());

    $message->save();

    $this->drupalGet($message->urlInfo()); /*May not be necessary*/

    // Simulate theming of the message.
    $view_builder = \Drupal::entityTypeManager()->getViewBuilder($message>getEntityTypeId());
    $output = \Drupal::service('renderer')->render($view_builder->view($message));

    $result = $this->xpath(); /*check if xpath returns a div on $output*/
    $this->assertEqual($result, $output); /*return TRUE / FALSE if xpath returns a div on $output*/
  }
```

RoySegall commented:
You could also use simple `strpos`

mccrodp committed:
Added new test for div wrapper element : testMessageTextWrapper()

mccrodp commented:
Need to fix this test, not sure where this pass by reference error is happening when I step through it.

Might try to use `strpos` instead of the xpath as I presume it has something to do with the call to setting raw render content or the xpath method.

mccrodp committed:
Fix for "passed by reference" error on MessageUiTest

mccrodp commented:
I kept the `xpath` method and fixed this by correcting the use of view builder. @RoySegall - I think this is ready...?

RoySegall commented on src/Tests/MessageUiTest.php:
```diff
@@ -119,6 +120,27 @@ public function testMessageTranslate() {
+    $this->setRawContent($output);
+
+    $xpath = $this->xpath('//div');
+    $this->assertTrue($xpath, 'A div has been found wrapping the message text.');
```
You can just do `strpos("<div>MESSAGE_TEXT", $output) !== FALSE`. Would be much more simple.

mccrodp commented:
Yes, I agree and this is also a better test as it forces a check on `MESSAGE_TEXT`. However, there is a deeper issue, that has been there for a while, with retrieving the message text during render that I will have to fix in the process\* :cry: I guess it needs to be fixed at some stage though.
- At the moment it is checking on `<div>\n \n</div>` as there is this message text issue.

RoySegall commented:
You don't need to get the message text from the render. you have the `getText()` method.

mccrodp commented:
Ahhh, yes, very true! Great, I'll try get this updated today and then look at the message text issue during render at a later time.

mccrodp commented:
So, I believe this will still have the same issue:

`strpos("<div>" . $message->getText(), $output) !== FALSE`

`$output` is from render not including the message text and `$message->getText()` will return the correct / expected message text and this operation will return `FALSE`.

I could just do `strpos("<div>", $output) !== FALSE` for the moment if that was acceptable?

mccrodp committed:
Change method to check for wrapper div from xpath to strpos in MessageUiTest.

RoySegall commented:
Hmm... probably. Let me see the most optional solution about this one. Either way i'll emerge this one tonight.

mccrodp commented:
Ok, I'll just get this green again and then leave it with you.

I wonder if it's related to `template_preprocess_message` and if the partials need to be explicitly loaded into `#markup` in there also. It's the only difference I can see from the render workflow compared to node.

mccrodp committed:
Remove whitespace in div element in template, attempt to fix MessageUiTest.

mccrodp committed:
Revert commit bfe566c - change wrapper detection back to xpath.

mccrodp commented:
@RoySegall - I reverted back to using xpath method as `string` property on the `$output` object is protected and the `__toString()` method throws an error message.

I think this is good to go as-is, unless you can improve this evening.

On a related note, in terms of the locating where the message text in `#markup` gets wiped, I've located it down to `\Drupal\Core\Render\Renderer::doRender()` line: 517 in this `Markup::create` method. I haven't gone through exactly why yet, but it exists before this call and afterwards it is overwritten with `\n \n`

`$elements['#markup'] = Markup::create($prefix . $elements['#children'] . $suffix);`

RoySegall commented:
Nope. Still not work as i expected. When trying to run 

``` php
    $message = Message::load(1);

    $build = \Drupal::entityTypeManager()->getViewBuilder('message')->view($message);
    $output = \Drupal::service('renderer')->renderPlain($build);
```

![adsadasd___site-install](https://cloud.githubusercontent.com/assets/1222368/12511135/a3c31ece-c117-11e5-9e8b-6db1f7db2c32.png)

And when trying to run 

``` php
\Drupal::theme()->render('message', ['elements' => ['#message' => $message]]);
```

I get 

> The website encountered an unexpected error. Please try again later.
> LogicException: Render context is empty, because render() was called outside of a renderRoot() or renderPlain() call. Use renderPlain()/renderRoot() or #lazy_builder/#pre_render instead. in Drupal\Core\Render\Renderer->doRender() (line 246 of core/lib/Drupal/Core/Render/Renderer.php).
> Drupal\Core\Render\Renderer->render(Array)
> Drupal\Core\Template\TwigExtension->escapeFilter(Object, Array, 'html', NULL, 1)
> __TwigTemplate_f0349ef89a49cf9dfa18a7a3a7d8cc090d647396c1d3f34cdf47385abcd122a2->doDisplay(Array, Array)
> Twig_Template->displayWithErrorHandling(Array, Array)
> Twig_Template->display(Array)
> Twig_Template->render(Array)
> twig_render_template('modules/message/templates/message.html.twig', Array)
> Drupal\Core\Theme\ThemeManager->render('message', Array)
> Drupal\devel\EventSubscriber\DevelEventSubscriber->onRequest(Object, 'kernel.request', Object)
> Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch('kernel.request', Object)
> Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1)
> Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1)
> Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1)
> Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1)
> Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1)
> Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1)
> Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1)
> Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1)
> Stack\StackedHttpKernel->handle(Object, 1, 1)
> Drupal\Core\DrupalKernel->handle(Object)

We need a way to render the message using API function(maybe theme or even apply a new method on the class) so this would help us in message ui and message notify and impact on message subscribe as well.

mccrodp commented:
@RoySegall Can you give me an example of what is expected on the first issue? 

Also, where are you running both of these code snippets from?

```
LogicException: Render context is empty, because render() was called outside of a renderRoot() or renderPlain() call.
```

This 2nd issue is as it says I believe, you are running a render out of context and if so you need to run renderRoot as I do in MessageUiTest in this PR.
E.g.

```
$build = \Drupal::entityTypeManager()->getViewBuilder('message')->view($message);
$output = \Drupal::service('renderer')->renderRoot($build);
```

RoySegall commented:
I'm running it from `DevelEventSubscriber::onRequest` which is the devel_init in D8. What I would expect is to run the code and get the render message content.

mccrodp commented:
Ok, not sure how to use that, but I'm sure I can figure it out. I think you probably have a valid issue with the 1st problem, so I'll take a look. 

The 2nd error is a valid error I believe and would also happen with any other entity when calling that render method out of context. So it seems anyways.

mccrodp commented:
I added the below to `DevelEventSubscriber::onRequest` directly.

```
$message = Message::load(2);

$build = \Drupal::entityTypeManager()->getViewBuilder('message')->view($message);
$output = \Drupal::service('renderer')->renderPlain($build);

dpm($output);
```

I could see your issue, there are 2 reasons:
1. As mentioned before, there is an issue with the display of the partials as explained previously and in https://github.com/Gizra/message/pull/48#issuecomment-173540603 We said we would fix this later
2. Fields need to be set to display, changing them from their hidden default at: `/admin/structure/message/manage/example_create_node/display`

I did then get the following template from `dpm()` as expected: https://gist.github.com/mccrodp/322835a0253b2ed01d33

RoySegall commented:
Enable the `kint` module.

mccrodp commented:
Ahh, thanks! Was wondering how to get pretty output. 

Won't be looking at this again for today. But I hope it can get merged and then look at any remaining issues after. As far as I'm aware they are not directly related to the pr code itself.

RoySegall commented:
This is the last issue before the stable release.

amitaibu commented:
Exciting :) but please don't tag a final release yet...

RoySegall commented:
Of course.

amitaibu commented:
:+:

amitaibu commented:
:+1:

mccrodp commented:
That sounds great. Is there anything left in this PR before it gets merged then?

Is the partials issue (https://github.com/Gizra/message/pull/48#issuecomment-173540603) fix is required before we tag a release? If so, should we open a separate PR, or what is best?

RoySegall commented:
> That sounds great. Is there anything left in this PR before it gets merged then?

Yes. This PR need to fix the issue with the message render. I don't see any change since last Friday, when I checked it. In this PR i need to be able to use the theme function to render the message in different view modes and get the text of the message.

mccrodp commented:
When you say "text of the message", you mean the partial? The fields are working fine for me, they just need to be enabled for display in the Manage Display.

I believe this is an issue separate to any code I have introduced in this PR, it's to do with the partials and render arrays in general, not any template related code. The `#markup` index of the build array is being overwritten as highlighted (by code / line no.) in https://github.com/Gizra/message/pull/48#issuecomment-173540603

This PR was originally just about making "...the template discoverable" and I believe it has done that. I won't get to look at this partials issue for about 2 weeks, so I recommend to merge this and open a new issue / PR for the partials render problem above. That means someone else can easily contribute a fix for this separately, your call! Thanks
